src/transform/src/analysis.rs

```diff
@@ -568,12 +577,81 @@ mod types {
+                MirRelationExpr::Get {
+                    id: mz_expr::Id::Local(i),
+                    typ,
+                    ..
+                } => {
+                    let mut result = typ.column_types.clone();
+                    if let Some(o) = depends.bindings().get(i) {
+                        if let Some(t) = results.get(*o) {
+                            if let Some(rec_typ) = t {
+                                // Reconcile nullability statements.
+                                // Unclear if we should trust `typ`.
+                                assert_eq!(result.len(), rec_typ.len());
+                                result.clone_from(rec_typ);
+                                for (res, col) in result.iter_mut().zip(typ.column_types.iter()) {
+                                    if !col.nullable {
+                                        res.nullable = false;
+                                    }
+                                }
+                            } else {
+                                // Our `None` information indicates that we are optimistically
+                                // assuming the best, including that all columns are non-null.
+                                // This should only happen in the first visit to a `Get` expr.
+                                // Use `typ`, but flatten nullability.
+                                for col in result.iter_mut() {
+                                    col.nullable = false;
+                                }
+                            }
+                        }
+                    }
+                    Some(result)
+                }
+                _ => {
+                    // Every expression with inputs should have non-`None` inputs at this point.
+                    let input_cols = offsets.into_iter().rev().map(|o| {
+                        o.as_ref()
+                            .expect("RelationType analysis discovered type-less expression")
+                    });
+                    Some(expr.try_col_with_input_cols(input_cols).unwrap())
+                }
+            }
+        }
+
+        fn lattice() -> Option<Box<dyn Lattice<Self::Value>>> {
+            Some(Box::new(RTLattice))
+        }
+    }
+
+    struct RTLattice;
+
+    impl Lattice<Option<Vec<ColumnType>>> for RTLattice {
+        fn top(&self) -> Option<Vec<ColumnType>> {
+            None
+        }
+        fn meet_assign(&self, a: &mut Option<Vec<ColumnType>>, b: Option<Vec<ColumnType>>) -> bool {
+            match (a, b) {
+                (_, None) => false,
+                (Some(a), Some(b)) => {
+                    let mut changed = false;
+                    assert_eq!(a.len(), b.len());
+                    for (at, bt) in a.iter_mut().zip(b.iter()) {
+                        assert_eq!(at.scalar_type, bt.scalar_type);
+                        if !at.nullable && bt.nullable {
+                            at.nullable = true;
```

Comment:
I suppose this is the moment where we set things nullable, undoing the optimism in the `Get` case above.